pkg/util/iptables/iptables.go

```diff
@@ -524,6 +481,73 @@ func (runner *runner) checkRuleUsingCheck(args []string) (bool, error) {
 	return false, fmt.Errorf("error checking rule: %v: %s", err, out)
 }
 
+const (
+	// Max time we wait for an iptables flush to complete after we notice it has started
+	iptablesFlushTimeout = 5 * time.Second
+	// How often we poll while waiting for an iptables flush to complete
+	iptablesFlushPollTime = 100 * time.Millisecond
+)
+
+// Monitor is part of Interface
+func (runner *runner) Monitor(reloadFunc func(), interval time.Duration, stopCh <-chan struct{}) {
+	bytes := make([]byte, 5)
+	rand.Read(bytes)
+	canaryChain := Chain("KUBE-CANARY-" + base32.StdEncoding.EncodeToString(bytes))
+
+	for {
+		// Create a "canary in the coal mine" whose deletion will let us know that
+		// iptables has been flushed. Because the iptables kernel API is terrible,
+		// even just checking that a chain exists can be very slow if the chain is
+		// in a table with very many rules. Thus, we use a chain in the "mangle"
+		// table as our primary canary, since that table ought to always be small
+		// (unlike "nat" and "filter").
+		if _, err := runner.EnsureChain(TableMangle, canaryChain); err != nil {
+			klog.Warningf("Could not set up iptables monitoring: %v", err)
+			return
+		}
+		_, _ = runner.EnsureChain(TableFilter, canaryChain)
+		_, _ = runner.EnsureChain(TableNAT, canaryChain)
+
+		// Poll until stopCh is closed or iptables is flushed
+		err := utilwait.PollUntil(interval, func() (bool, error) {
+			// FlushChain will succeed if the chain exists and fail if it doesn't.
+			// (And the chain is empty anyway so it won't have any other effect.)
+			if err := runner.FlushChain(TableMangle, canaryChain); err == nil {
+				return false, nil
+			}
+
+			// Our canary in the "mangle" table has been deleted; wait for the
+			// other canaries to be deleted too before returning so we don't
+			// start reloading too soon.
+			err := utilwait.PollImmediate(iptablesFlushPollTime, iptablesFlushTimeout, func() (bool, error) {
```

Comment:
How would the caller know what a good timeout and period are though? How can the caller, in isolation, know what a good "low-impact" table is, given that this depends on what rules *other people* are creating? What the caller wants is just "if iptables gets flushed, call me", and `utiliptables` should be responsible for figuring out the best way to implement that for them.

So of the parameters:
- polling interval: I made this an argument to `Monitor` mostly because if I didn't then kube-proxy's `iptablesSyncPeriod` would become a no-op. But still, I think unlike the other stuff, this *is* something the caller might care about. Different callers might have different tolerances for how long they're willing to be out of sync.
- trigger table (`"mangle"`) - This needs to be decided based on the system as a whole. We could autodetect a good table by scanning `iptables-save` output, or we could add kubelet and kube-proxy config arguments to let the admin configure it based on their knowledge of their system. But mostly, the problem we're trying to work around is that kubernetes itself makes the `nat` and `filter` tables get huge, and I haven't heard of any systems where that's also a problem with `mangle`, `raw`, or `security`. I picked `mangle` somewhat at random from the three. (`security` seemed like a bad idea.) If we always used `mangle` now and that turned out to be a problem we could do the autodetect thing later.
- canary chain name (`KUBE-CANARY-XXXXXX`) - We don't let people rename any of our other chains... I guess maybe you're arguing that `pkg/util/iptables` is too low-level to be imposing the `KUBE-` prefix on the caller, and maybe that's true...
- post-trigger-table-flush poll interval (`100ms`) and timeout (`5s`) - I totally pulled these numbers out of my ass. I'm happy to revisit them but I don't think we need to make them configurable. All callers want the same behavior here ("respond quickly; don't get stuck if something goes wrong").

  FTR, basically all of the firewall managers do the same thing when reloading:

      for table in $(cat /proc/net/ip_tables_names); do
          iptables -F -t $table
          iptables -X -t $table
      done

  So we're just trying to wait for that loop to complete, which normally should happen very quickly, but maybe sometimes might take a bit longer, but shouldn't ever take *very* long.
- other tables to wait for (`["nat", "filter"]`) - OK, you got me, I am totally cheating there. But we could fix this without the caller having to pass any more information; the `runner` already knows what tables the user cares about based on the `EnsureChain`/`EnsureRule`/`Restore` commands they have run, so we can just figure out what tables to wait for based on that.

Reply:
> trigger table (`"mangle"`) - This needs to be decided based on the system as a whole. 
> We could autodetect a good table by scanning `iptables-save` output, or we could add 
> kubelet and kube-proxy config arguments to let the admin configure it based on their 
> knowledge of their system. But mostly, the problem we're trying to work around is that 
> kubernetes itself makes the `nat` and `filter` tables get huge, and I haven't heard of any 
> systems where that's also a problem with `mangle`, `raw`, or `security`. I picked `mangle` 
> somewhat at random from the three. (`security` seemed like a bad idea.) If we always used 
> `mangle` now and that turned out to be a problem we could do the autodetect thing later.

My thinking here was that the knowledge that "kubernetes itself makes the `nat` and `filter` tables get huge" doesn't belong in a nearly-general-purpose lib like this.  I could easily see moving this lib to a new repo k/iptables or just k/utils/iptables.  At least kube-proxy and kubelet know that kubernetes (ab)uses nat and filter, so they are in a position to give a hint.  Maybe the hint is "use mangle" or "use anything except nat or filter" instead? 

Auto-detecting sounds good, but I worry a little about point-in-time samples (e.g. at machine boot time, nat might be a good choice, because kube-proxy hasn't spun up yet.  So I'd still say we need a hint, at least.  "not nat, not filter."

> canary chain name (`KUBE-CANARY-XXXXXX`) - We don't let people rename any of our 
> other chains... I guess maybe you're arguing that `pkg/util/iptables` is too low-level to be 
> imposing the `KUBE-` prefix on the caller, and maybe that's true...

Yes, I don't think this should have "KUBE" coded in, and I assume the rand is because you can have multiple agents Monitor()ing, but I thin rand is kind of busted (as I said elsewhere).

The rest I don't care so much :)

> other tables to wait for (`["nat", "filter"]`) - OK, you got me, I am totally cheating there. But we 
> could fix this without the caller having to pass any more information; the `runner` already 
> knows what tables the user cares about based on the `EnsureChain`/`EnsureRule`/`Restore` > commands they have run, so we can just figure out what tables to wait for based on that.

I would expect Monitor() to be called very early, so kube-proxy might not have called any Ensure yet.  Parsing Restore blobs seems clunky to me.  It could be that the "blockers list" *is* the canary-table hint.  Don't canary in blocker tables.